File: 173-binary-search-tree-iterator/173-binary-search-tree-iterator.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BSTIterator {
    private Stack<TreeNode> st;
    
    private void pushAll(TreeNode root){
        
        while(root != null){
            st.push(root);
            root = root.left;
        }
    }
    
    public BSTIterator(TreeNode root) {
        st = new Stack();
        pushAll(root);
    }
    
    public int next() {
        
        TreeNode tmpNxt = st.peek();
        st.pop();
        
        pushAll(tmpNxt.right);
        return tmpNxt.val;
    }
    
    public boolean hasNext() {
        return st.size() != 0;
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */

File: 173-binary-search-tree-iterator/173-binary-search-tree-iterator.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTIterator {
private:
    stack<TreeNode*> myStack;
    
    void pushAll(TreeNode* root){
        
        TreeNode* node = root;
        while(node != NULL){
            myStack.push(node);
            node = node->left;
        }
    }
    
public:
    BSTIterator(TreeNode* root) {
        pushAll(root);
    }
    
    int next() {
        
        TreeNode* nextVal = myStack.top();
        myStack.pop();
        pushAll(nextVal->right);
        return nextVal->val;
    }
    
    bool hasNext() {
        if(myStack.size()>0){
            return true;
        }
        else{
            return false;
        }
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */